[libs] Make Control-Plane API Bi-directional

csegarragonz opened the pull request:
In this PR I enhance the control-plane API to support more complex messages, by moving to `bincode`-powered (de)-serialization, and supporting replies.

For the time-being, the control-plane API is bi-directional from nanvixd to the system/user VM and vice-versa. When the message is sent from nanvixd, we don't differentiate between recipients.

In follow-up PRs I will add messages from the system VM to nanvixd, to help with synchronization during system provisioning. 

We can also explore merging the user-vm API with this control-plane API. In this case, however, we should also consider having a control-plane stream between the system and the user VM. If I am not mistaken this feature is also desirable for snapshots.

csegarragonz committed:
[libs] E: Make Control-Plane API Bi-Directional

In this commit I extend the control-plane-api library to support
general-purpose, bi-directional messages from nanvixd to the system/user
vm.

In follow-up PRs I will implement the messages from the system and the
user vm, respectively.

Having a more expressive control-plane library will help reducing race
conditions during system provisioning time.

Signed-off-by: Carlos Segarra <carlos@carlossegarra.com>

csegarragonz committed:
[libs] E: Add MAX_MESSAGE_LEN Constant To Syscomm

Signed-off-by: Carlos Segarra <carlos@carlossegarra.com>

csegarragonz committed:
[utils] B: Refactor Nanvixd To Use New Control-Plane API

Signed-off-by: Carlos Segarra <carlos@carlossegarra.com>

csegarragonz committed:
[daemons] B: Refactor Linuxd To Use New Control-Plane API

Signed-off-by: Carlos Segarra <carlos@carlossegarra.com>

csegarragonz committed:
[microvm] B: Refactor Microvm To Use New Control-Plane API

Signed-off-by: Carlos Segarra <carlos@carlossegarra.com>

csegarragonz commented on src/libs/control-plane-api/src/lib.rs:
```diff
@@ -17,32 +27,150 @@ use ::syscomm::{
+    let mut length_bytes: [u8; 4] = [0u8; 4];
+    let num_read: usize = stream.try_read_exact(&mut length_bytes)?;
+    debug_assert_eq!(num_read, 4);
+    let len: usize = u32::from_le_bytes(length_bytes) as usize;
```
I think it is safe to assume that `size_of::u32` <= `size_of::usize`?

ppenna replied:
Yes, it is. Please add the following (if possible) to `lib.rs`:

```
#![forbid(clippy::cast_possible_truncation)]
#![forbid(clippy::cast_possible_wrap)]
#![forbid(clippy::cast_precision_loss)]
#![forbid(clippy::unnecessary_cast)]
#![forbid(invalid_reference_casting)]
#![forbid(clippy::fn_to_numeric_cast)]
#![forbid(clippy::fn_to_numeric_cast_with_truncation)]
```

csegarragonz committed:
[utils] B: Refactor Nanvix-Bench To Use New Control-Plane API

Signed-off-by: Carlos Segarra <carlos@carlossegarra.com>

ppenna commented on src/libs/config/src/lib.rs:
```diff
@@ -54,13 +54,21 @@ pub mod system {
 
 pub mod syscomm {
 
+    ///
+    /// # Description
+    ///
+    /// Provides the maximum length of a message we receive. We set a limit to
+    /// prevent a malformed payload to trigger unbounded allocations in linuxd.
+    ///
+    pub const MAX_MESSAGE_LEN: usize = 128 * 1024 * 1024;
+
```
Why is this new constant required?

csegarragonz replied:
It is used as a sanity-check before we allocate a buffer for the newly received message: https://github.com/nanvix/nanvix/blob/enhancement-control-plane-api-bidirectional/src/libs/control-plane-api/src/lib.rs#L159-L164

(We read the message length from the payload itself as part of the wire-format.)

ppenna replied:
Sure, but why could't we use directly `MAX_LINUXD_MESSAGE_LEN`, or rename `MAX_LINUXD_MESSAGE_LEN` to `MAX_MESSAGE_LEN` ?

csegarragonz replied:
Its not exactly a LINUXD message, so that is why I added a new constant. I did not want to refactor MAX_LINUXD in this PR to not add more noise but, if you are happy about it, I can refactor it in this PR.

csegarragonz replied:
Done!

csegarragonz commented on src/libs/control-plane-api/src/lib.rs:
```diff
@@ -17,32 +27,156 @@ use ::syscomm::{
+///
+/// In case of success, returns nothing. Otherwise, returns an error.
+///
+pub fn send_command<T: WireCommand>(stream: &mut SocketStream, msg: &T) -> Result<(), SocketError> {
+    let payload: Vec<u8> = bincode::serde::encode_to_vec(msg, bincode_cfg())
```
Messages may get more complicated down the road, so I'd stick to passing by reference.

csegarragonz commented on src/libs/control-plane-api/src/lib.rs:
```diff
@@ -17,32 +27,156 @@ use ::syscomm::{
+        })?;
+
+
+    if payload.len() > u32::MAX as usize {
+        let reason: String = format!("payload length exceeds maximum allowed size (length={}, max={})", payload.len(), u32::MAX);
+        error!("{reason}");
+        return Err(Error::new(ErrorKind::InvalidData, reason).into());
+    }
```
I think we do need this check, and you suggested it yourself before :D

ppenna commented on src/libs/control-plane-api/src/lib.rs:
```diff
@@ -17,32 +35,162 @@ use ::syscomm::{
+/// Default bincode options to encode/decode control-plane messages. This configuration uses
+/// fixed-size ints, and little-endian encoding.
+///
+fn bincode_cfg() -> impl bincode::config::Config {
```
This can be made a `const` function I believe.

ppenna reviewed (dismissed):
I understand that control messages may get more complex down the road, still using 'serdes' looks a bit overkill now

csegarragonz committed:
[utils] B: Add Timeout In Nanvixd's Sandbox Clean-Up

In this commit I patch nanvixd to take a timeout when waiting for the
graceful shutdown of the system or user VM.

It is good that we try and cleanly shutdown the different components,
but if something goes wrong nanvixd cannot wait and hang forever.
Instead, if we hit a timeout we send a SIGKILL to the child.

Signed-off-by: Carlos Segarra <carlos@carlossegarra.com>

csegarragonz commented:
> I understand that control messages may get more complex down the road, still using 'serdes' looks a bit overkill now

I have a PR where I am adding the user vm ID in a message (a `u32`) and down the road I want to merge the control-plane API with the user-vm API (which also uses `serdes`). 

I am open to re-assessing in the future and moving to something more lightweight.

csegarragonz committed:
[libs] E: Remove MAX_LINUXD_MESSAGE_LEN Constant

Signed-off-by: Carlos Segarra <carlos@carlossegarra.com>
